File: lexer.go
package main

import (
    "slices"
)

type Token struct {
    Type TokenType
    Value string
}

type TokenType int

const (
    ID TokenType = iota
    Flow
    Op
    Number
    String
)

func lex(data string) []Token {
    chars := []string{" ", "=", "{", "}", "(", ")", "[", "]", "\n"}
    ops := []string{"+", "!", "~", "%", "*", "&", "/", ":", "-", "|", "^"}
    numbers := []string{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"}

    tokens := []Token{}
    tok := ""
    comment := false
    str := ' '
    
    consume := func(Type TokenType){
        number := true
        if tok == " " || tok == "\n" {
            tok = ""
        }
        for _, n := range tok {
            if !slices.Contains(numbers, string(n)){
                number = false
            }
        }
        if number { Type = Number }
        if len(tok) > 0 {
            tokens = append(tokens, Token{Type, tok})
            tok = ""
        }
    }

    for _, token := range data {
        st := string(token)

        if comment {
            if token == '\n' {
                comment = false
            } else if token == '#' {
                comment = false
            }
            tok = ""
        } else if token == '#' {
            consume(ID)
            comment = true
        } else if str != ' ' {
            tok += st
            if token == str {
                str = ' '
                consume(String)
            }
        } else if token == '"' || token == '\'' {
            consume(ID)
            str = token
            tok = st
        } else if slices.Contains(chars, st) {
            consume(ID)
            tok = st
            consume(Flow)
        } else if slices.Contains(ops, st) {       
            consume(ID)
            tok = st
            consume(Op)
        } else {
            tok += st
        }
    }

    consume(ID)

    return tokens
}

File: lexer.js
class Token {
    constructor(t, v, l, c){
        this.type = t
        this.value = v
        this.line = l
        this.ch = c
    }
}

function lex(data){
    tokens = []
    line = 1
    ch = 1

    flow = ["=", "{", "}", "(", ")", "[", "]", ":"]
    ops = ["+", "!", "~", "%", "*", "&", "/", "-", "|", "^"]
    numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]

    comment = false
    string = ""
    tok = ""

    newline = function(){
        line++
        ch = 1
        tok = ""
    }

    consume = function(type){
        number = true
        for (i in tok){
            if(!numbers.includes(tok[i])){
                number = false
            }
        }
        type = number ? "number" : type
        if (tok.length > 0){
            tokens.push(new Token(type, tok, line, ch))
        }
        ch += tok.length
        tok = ""
    }

    for(i in data){
        cur = data[i]
        if(comment){
            if(cur == "\n"){
                newline()
                comment = false
            } else if (cur == "#"){
                comment = false
            }
            ch++
            tok = ""
        } else if(string){
            if(cur == string){
                string = ""
                consume("string")
                ch++
            } else {
                tok += cur
            }
        } else if(cur == " "){
            consume("id")
            ch++
        } else if(cur == "#"){
            consume("id")
            comment = true
        } else if(cur == "\"" || cur == "'"){
            consume("id")
            string = cur
            ch++
        } else if(cur == "\n"){
            consume("id")
            newline()
        } else if(flow.includes(cur)){
            consume("id")
            tok = cur
            consume(cur)
        } else if(ops.includes(cur)){
            consume("id")
            tok = cur
            consume("op")
        } else {
            tok += cur
        }
    }
    consume("id")

    return tokens
}

module.exports = lex
